### PR TITLE
modules: hal: nordic: Add missing mappings of peripheral base addresses

### PR DESCRIPTION
**Pull request by anangl**
For quite a few peripherals that are currently supported by nrfx HALs
or drivers there are no definitions of corresponding CMSIS-Core
peripheral accessing symbols that would provide their base addresses
in the proper domain (secure or non-secure), accordingly to the build
target. This commits adds devicetree nodes for these peripherals and
updates the revision of the hal_nordic module so that it can provide
the missing symbol definitions basing on data from the devicetree.

---

Requires https://github.com/zephyrproject-rtos/hal_nordic/pull/3 to go in first.

### CODE REVIEW

**Comment by ulfalizer** [comment]
@anangl 
Might want to get this in via the [topic-dts](https://github.com/zephyrproject-rtos/zephyr/tree/topic-dts) branch and change

```yaml
inherits:
    !include base.yaml
```

to

```yaml
include: base.yaml
```

and

```
category: optional/required
```

to

```
required: false/true
```

that'll save some later changes, though the stuff on the `topic-dts` branch is mostly backwards-compatible.

**Review by ioannisg — COMMENTED** [review]
It's more like "dts: nordic: nrf9160: " It seems to be a PR for DTS updates for nRF9160

**Comment by anangl** [comment]
I forgot about the aliases for non-secure part. As usual...
Fixed.

**Comment by anangl** [comment]
@ioannisg 
> It's more like "dts: nordic: nrf9160: " It seems to be a PR for DTS updates for nRF9160

The actual purpose of these changes in dts is to provide data for these mappings of base addresses. That's why I used such title. But indeed, it seems that it will be clearer to use two commits - one with the dts changes and one with the update of west manifest. What you think?

**Comment by ioannisg** [comment]
> 
> 
> @ioannisg
> 
> > It's more like "dts: nordic: nrf9160: " It seems to be a PR for DTS updates for nRF9160
> 
> The actual purpose of these changes in dts is to provide data for these mappings of base addresses. That's why I used such title. But indeed, it seems that it will be clearer to use two commits - one with the dts changes and one with the update of west manifest. What you think?

Yeap, that seems better

**Review by pabigot — COMMENTED** [review]
This looks fine as a cleanup, though I second @ulfalizer's recommendation to wait until material currently staged in topic-dts is merged.  Also waiting until west.yml references a SHA1 before approving.

One thing I don't understand, which this PR makes very obvious, is why there aliases (e.g. `i2c-0`) for every node label (`i2c0`) (not the same as the node `label` property value "I2C_0").  There are a very small number of situations where an alias is still necessary, primarily for cross-target common resources like `sw0` and `led1`, which probably should be handled a different way.

In all of Zephyr master there is one driver, three samples, and four tests that reference generated alias symbols.

It's not in scope for this PR, but we should move towards getting rid of those.

**Comment by anangl** [comment]
I changed the target branch to `topic-dts` and applied the necessary corrections in bindings, as @ulfalizer suggested. Sorry for this noise in between, I haven't noticed that I initially used some old state of the `topic-dts` branch.
 I also split the changes into two commits, as agreed with @ioannisg.

**Comment by anangl** [comment]
@pabigot 
> One thing I don't understand, which this PR makes very obvious, is why there aliases (e.g. `i2c-0`) for every node label (`i2c0`) (not the same as the node `label` property value "I2C_0"). There are a very small number of situations where an alias is still necessary, primarily for cross-target common resources like `sw0` and `led1`, which probably should be handled a different way.
> 
> In all of Zephyr master there is one driver, three samples, and four tests that reference generated alias symbols.
> 
> It's not in scope for this PR, but we should move towards getting rid of those.

We use aliases to get macros generated from dts without base addresses in their names. We use such macros in drivers and also in _nrfx_config_nrf9160.h_. The thing is that the `DT_INST_*` macros cannot be used for these purposes as the index contained in the macro name does not necessarily match the index of the hardware instance it refers to. For instance `DT_INST_0_NORDIC_NRF_I2C_BASE_ADDRESS` may not refer to `NRF_TWIM0` while `DT_NORDIC_NRF_I2C_I2C_0_BASE_ADDRESS`, which is generated for the `i2c-0` alias, always refers to this particular hardware instance.

File: dts/arm/nordic/nrf9160.dtsi [hunk 1]
```diff
@@ -37,12 +37,20 @@
 		uart-2 = &uart2;
 		uart-3 = &uart3;
 		adc-0 = &adc;
+		egu-0 = &egu0;
```
Comment (galak) [inline dts/arm/nordic/nrf9160.dtsi]: Do we really need to add these aliases?  Can we not use DT_INST defines?

Comment (anangl) [reply dts/arm/nordic/nrf9160.dtsi]: No, we can't use `DT_INST_*` macros, see my [comment](https://github.com/zephyrproject-rtos/zephyr/pull/18937#issuecomment-528838585) above.

**Comment by pabigot** [comment]
OK, I see the use.  Take this example:
```
#define DT_NORDIC_NRF_SAADC_40007000_BASE_ADDRESS   0x40007000
#define DT_ALIAS_ADC_0_BASE_ADDRESS                 DT_NORDIC_NRF_SAADC_40007000_BASE_ADDRESS
#define DT_NORDIC_NRF_SAADC_ADC_0_BASE_ADDRESS      DT_NORDIC_NRF_SAADC_40007000_BASE_ADDRESS
#define DT_INST_0_NORDIC_NRF_SAADC_BASE_ADDRESS     DT_NORDIC_NRF_SAADC_40007000_BASE_ADDRESS
```
which derives from 

                adc: adc@40007000 {
                        compatible = "nordic,nrf-saadc";
                        reg = <0x40007000 0x1000>;
                        interrupts = <7 1>;
                        status = "disabled";
                        label = "ADC_0";
                };

plus:

        aliases {
                adc-0 = &adc;
        };

I would like to eliminate the alias and instead get something like:

```
#define DT_NORDIC_NRF_SAADC_40007000_BASE_ADDRESS   0x40007000
#define DT_LABEL_ADC_BASE_ADDRESS                   DT_NORDIC_NRF_SAADC_40007000_BASE_ADDRESS
#define DT_INST_0_NORDIC_NRF_SAADC_BASE_ADDRESS     DT_NORDIC_NRF_SAADC_40007000_BASE_ADDRESS
```
where in the case of a labelled node we get something that allows code to reference using the label directly, rather than through an alias that adds no value.  That doesn't mean we can't keep the `DT_ALIAS_` prefix just as it is when there is an alias, but the label on the node is generally all we need.

I can't tell whether `DT_NORDIC_NRF_SAADC_ADC_0_BASE_ADDRESS` came from the alias `adc-0` or the label property `"ADC_0"`, but in either case it should be unnecessary too.

@galak should I add this as a devicetree issue?

**Comment by anangl** [comment]
> I would like to eliminate the alias and instead get something like:
> 
> ```
> #define DT_NORDIC_NRF_SAADC_40007000_BASE_ADDRESS   0x40007000
> #define DT_LABEL_ADC_BASE_ADDRESS                   DT_NORDIC_NRF_SAADC_40007000_BASE_ADDRESS
> #define DT_INST_0_NORDIC_NRF_SAADC_BASE_ADDRESS     DT_NORDIC_NRF_SAADC_40007000_BASE_ADDRESS
> ```
> 
> where in the case of a labelled node we get something that allows code to reference using the label directly, rather than through an alias that adds no value.

Looks good to me. I don't like these aliases, too. And I obviously would be glad if we could get rid of them.

**Comment by anangl** [comment]
Rebased on master.

**Comment by anangl** [comment]
Updated _west.yml_ with the proper revision of the hal_nordic module as https://github.com/zephyrproject-rtos/hal_nordic/pull/3 has been merged.

**Review by pabigot — APPROVED** [review]

File: dts/arm/nordic/nrf9160.dtsi (hunk 1 above)
Comment (galak) [reply dts/arm/nordic/nrf9160.dtsi]: Than, lets not add new aliases.

Comment (galak) [reply dts/arm/nordic/nrf9160.dtsi]: Sorry, misread your comment.

Comment (pabigot) [reply dts/arm/nordic/nrf9160.dtsi]: If #18986 gets approved for work and resolved then we can remove them.

**Comment by galak** [comment]
> We use aliases to get macros generated from dts without base addresses in their names. We use such macros in drivers and also in _nrfx_config_nrf9160.h_. The thing is that the `DT_INST_*` macros cannot be used for these purposes as the index contained in the macro name does not necessarily match the index of the hardware instance it refers to. For instance `DT_INST_0_NORDIC_NRF_I2C_BASE_ADDRESS` may not refer to `NRF_TWIM0` while `DT_NORDIC_NRF_I2C_I2C_0_BASE_ADDRESS`, which is generated for the `i2c-0` alias, always refers to this particular hardware instance.

What's the issue with the `_INST_` defines?  I understand that INST_0 doesn't necessary refer to the same HW instance always, but when do you need to know which specific hw instance it is?

**Comment by pabigot** [comment]
@galak The hardware address for the instance is being taken from the devicetree information.

**Comment by galak** [comment]
> @galak The hardware address for the instance is being taken from the devicetree information.

what do you mean by the hardware address?  If it's taken from the DT info, then why do we care about needing to know the specific instance from the HW perspective?

The reason DT_INST_ can't be used for various drivers on most SoCs is related to pinmux info.  Once we're able to generate pinmux info for a given SoC that should free up the driver to utilize DT_INST_ defines.  So I'm trying to understand in the nordic case what prevents to the use of DT_INST_ defines today.

**Comment by pabigot** [comment]
This isn't my area, but: what @nvlsianpu was referencing above is that Nordic HAL module in its Zephyr fork uses the devicetree content to determine the hardware addresses to use for symbols like `PWM0` which are supposed to correspond to the first instance of the PWM peripheral, rather than hard-coding a constant that might end up inconsistent with what's in the devicetree.

If Zephyr's identification of that instance is not the same as the one selected by other systems that Zephyr applications must coordinate with (such as a softdevice blob) things will go very badly wrong.  Because `DT_INST_0_FOO` is not always the vendor-defined `FOO0` you can't use the generated symbols that way.  You could use `DT_ALIAS_FOO_0` as now, or when supported avoid an alias and use `DT_LABEL_FOO_0` which is fixed to identify the node with a particular address.

**Comment by galak** [comment]
> This isn't my area, but: what ~@nvlsianpu~ sorry, @anangl was referencing above is that Nordic HAL module in its Zephyr fork uses the devicetree content to determine the hardware addresses to use for symbols like `PWM0` which are supposed to correspond to the first instance of the PWM peripheral, rather than hard-coding a constant that might end up inconsistent with what's in the devicetree.
> 
> If Zephyr's identification of that instance is not the same as the one selected by other systems that Zephyr applications must coordinate with (such as a softdevice blob) things will go very badly wrong. Because `DT_INST_0_FOO` is not always the vendor-defined `FOO0` you can't use the generated symbols that way. You could use `DT_ALIAS_FOO_0` as now, or when supported avoid an alias and use `DT_LABEL_FOO0` which is fixed to identify the node with a particular address.

@anangl where is this mapping happing between the `DT_ALIAS` define and nordic internals?

**Comment by pabigot** [comment]
https://github.com/zephyrproject-rtos/hal_nordic/blob/master/nrfx_config_nrf9160.h

**Comment by pabigot** [comment]
Or maybe not, as I don't see it there.  I know it was somewhere....  I'm going to stop with this discussion because it's clearly outside my ken, but I still want #18986 to get some love regardless.

**Comment by anangl** [comment]
> @anangl where is this mapping happing between the `DT_ALIAS` define and nordic internals?

These mappings do not use `DT_ALIAS` macros but the older ones generated for aliases: `DT_<compatible>_<alias_name>_*`. And they are in the file pointed out by @pabigot. [Here](https://github.com/zephyrproject-rtos/hal_nordic/blob/d8217cb96406172a1cd4493ddfa968cb167665f9/nrfx_config_nrf9160.h#L81) is one example from the revision of this file currently used by master (this PR updates the hal_nordic module to the revision that adds more such mappings).

**Comment by anangl** [comment]
Rebased on master and in result skipped the commit modifying _west.yml_ as the needed revision update already went in through #18895.

**Review by ioannisg — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

File: dts/bindings/arm/nordic,nrf-kmu.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+title: Nordic KMU
+
+description: >
+    Binding for the Nordic KMU (Key Management Unit)
+
+include: base.yaml
+
+properties:
+    compatible:
+      constraint: "nordic,nrf-kmu"
```
Comment (ulfalizer) [inline dts/bindings/arm/nordic,nrf-kmu.yaml]: Ditto here.

File: dts/bindings/arm/nordic,nrf-egu.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: Apache-2.0
+#
```
Comment (ulfalizer) [inline dts/bindings/arm/nordic,nrf-egu.yaml]: Could shorten the header a bit to

```yaml
# Copyright (c) 2019 Nordic Semiconductor ASA
# SPDX-License-Identifier: Apache-2.0
```

I changed it to that in a bunch of other files.

File: dts/bindings/arm/nordic,nrf-kmu.yaml [hunk 4]
```diff
@@ -0,0 +1,22 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: Apache-2.0
+#
```
Comment (ulfalizer) [inline dts/bindings/arm/nordic,nrf-kmu.yaml]: Ditto here re. header.

File: dts/bindings/audio/nordic,nrf-pdm.yaml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+title: Nordic PDM
+
+description: >
+    Binding for the Nordic PDM (pulse density modulation interface)
+
+include: base.yaml
+
+properties:
+    compatible:
+      constraint: "nordic,nrf-pdm"
```
Comment (ulfalizer) [inline dts/bindings/audio/nordic,nrf-pdm.yaml]: Ditto here re. `compatible`.

File: dts/bindings/audio/nordic,nrf-pdm.yaml [hunk 6]
```diff
@@ -0,0 +1,32 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: Apache-2.0
+#
```
Comment (ulfalizer) [inline dts/bindings/audio/nordic,nrf-pdm.yaml]: Ditto here re. header.

File: dts/bindings/i2c/nordic,nrf-twis.yaml [hunk 7]
```diff
@@ -0,0 +1,42 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: Apache-2.0
+#
```
Comment (ulfalizer) [inline dts/bindings/i2c/nordic,nrf-twis.yaml]: Ditto here re. header.

File: dts/bindings/i2c/nordic,nrf-twis.yaml [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+title: Nordic TWIS
+
+description: >
+    Binding for the Nordic TWIS (TWI slave with EasyDMA)
+
+include: base.yaml
+
+properties:
+    compatible:
+      constraint: "nordic,nrf-twis"
```
Comment (ulfalizer) [inline dts/bindings/i2c/nordic,nrf-twis.yaml]: Ditto here re. `compatible`.

File: dts/bindings/i2s/nordic,nrf-i2s.yaml [hunk 9]
```diff
@@ -0,0 +1,47 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: Apache-2.0
+#
```
Comment (ulfalizer) [inline dts/bindings/i2s/nordic,nrf-i2s.yaml]: Ditto here re. header.

File: dts/bindings/i2s/nordic,nrf-i2s.yaml [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+title: Nordic I2S
+
+description: >
+    Binding for the Nordic I2S (Inter-IC sound interface)
+
+include: i2s-controller.yaml
+
+properties:
+    compatible:
+      constraint: "nordic,nrf-i2s"
```
Comment (ulfalizer) [inline dts/bindings/i2s/nordic,nrf-i2s.yaml]: Ditto here re. `compatible`.

File: dts/bindings/power/nordic,nrf-regulators.yaml [hunk 11]
```diff
@@ -0,0 +1,19 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: Apache-2.0
+#
```
Comment (ulfalizer) [inline dts/bindings/power/nordic,nrf-regulators.yaml]: Ditto here re. header.

File: dts/bindings/power/nordic,nrf-regulators.yaml [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+title: Nordic REGULATORS
+
+description: >
+    Binding for the Nordic REGULATORS (voltage regulators control module)
+
+include: base.yaml
+
+properties:
+    compatible:
+      constraint: "nordic,nrf-regulators"
```
Comment (ulfalizer) [inline dts/bindings/power/nordic,nrf-regulators.yaml]: Ditto here re. compatible.

File: dts/bindings/power/nordic,nrf-vmc.yaml [hunk 13]
```diff
@@ -0,0 +1,19 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: Apache-2.0
+#
```
Comment (ulfalizer) [inline dts/bindings/power/nordic,nrf-vmc.yaml]: Ditto here re. header.

File: dts/bindings/power/nordic,nrf-vmc.yaml [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+title: Nordic VMC
+
+description: >
+    Binding for the Nordic VMC (Volatile Memory Controller)
+
+include: base.yaml
+
+properties:
+    compatible:
+      constraint: "nordic,nrf-vmc"
```
Comment (ulfalizer) [inline dts/bindings/power/nordic,nrf-vmc.yaml]: Ditto here re. `compatible`.

**Review by ulfalizer — CHANGES_REQUESTED** [review]
Some nits re. new binding stuff.

**Comment by ulfalizer** [comment]
edtlib prints deprecation warnings that say what needs to be changed btw.

**Comment by anangl** [comment]
@ulfalizer I addressed your comments.

**Review by ulfalizer — APPROVED** [review]
Thanks

**Pull request merged**